client/asset/btc/btc.go

```diff
@@ -2365,6 +2377,209 @@ func (btc *baseWallet) fund(keep uint64, minConfs uint32, lockUnspents bool,
+	} else {
+		splitTxSizeWithoutInputs += uint64(dexbtc.P2PKHOutputSize * numOutputs)
+	}
+	enough := func(inputSize, sum uint64) (bool, uint64) {
+		splitTxFee := (splitTxSizeWithoutInputs + inputSize) * splitTxFeeRate
+		req := totalOutputRequired + splitTxFee
+		return sum >= req, sum - req
+	}
+	fundSplitCoins, _, _, _, _, totalIn, err := btc.fundInternal(keep, 0, true, enough)
+	if err != nil {
+		return nil, nil, 0, err
+	}
+
+	baseTx, _, _, err := btc.fundedTx(fundSplitCoins)
+	if err != nil {
+		return nil, nil, 0, err
+	}
+
+	outputAddresses := make([]btcutil.Address, len(values))
+	for i, req := range requiredForOrders {
+		changeAddr, err := btc.node.changeAddress()
+		if err != nil {
+			return nil, nil, 0, err
+		}
+		outputAddresses[i] = changeAddr
+		script, err := txscript.PayToAddrScript(changeAddr)
+		if err != nil {
+			return nil, nil, 0, err
+		}
+		baseTx.AddTxOut(wire.NewTxOut(int64(req), script))
+	}
+
+	changeAddr, err := btc.node.changeAddress()
+	if err != nil {
+		return nil, nil, 0, err
+	}
+	tx, err := btc.sendWithReturn(baseTx, changeAddr, totalIn, totalOutputRequired, splitTxFeeRate)
+	if err != nil {
+		return nil, nil, 0, err
+	}
+	txHash := tx.TxHash()
+
+	coins := make([]asset.Coins, len(values))
+	redeemScripts := make([][]dex.Bytes, len(values))
+	ops := make([]*output, len(values))
+	for i := range coins {
+		coins[i] = asset.Coins{newOutput(&txHash, uint32(i), uint64(tx.TxOut[i].Value))}
+		redeemScripts[i] = []dex.Bytes{tx.TxOut[i].PkScript}
+		ops[i] = newOutput(&txHash, uint32(i), uint64(tx.TxOut[i].Value))
+		btc.fundingCoins[ops[i].pt] = &utxo{
+			txHash:  &txHash,
+			vout:    uint32(i),
+			amount:  uint64(tx.TxOut[i].Value),
+			address: outputAddresses[i].String(),
+		}
+	}
+	btc.node.lockUnspent(false, ops)
+
+	var totalOut uint64
+	for _, txOut := range tx.TxOut {
+		totalOut += uint64(txOut.Value)
+	}
+
+	return coins, redeemScripts, totalIn - totalOut, nil
+}
+
+// fundMulti first attempts to fund each of the orders with with the available
+// UTXOs. If a split is not allowed, it will fund the orders that it was able
+// to fund. If splitting is allowed, a split transaction will be created to fund
+// all of the orders.
+func (btc *baseWallet) fundMulti(keep uint64, values []*asset.MultiOrderValue, splitTxFeeRate, maxFeeRate uint64, allowSplit bool, splitBuffer uint64) ([]asset.Coins, [][]dex.Bytes, uint64, error) {
+	btc.fundingMtx.Lock()
+	defer btc.fundingMtx.Unlock()
+
+	reserves := btc.reserves()
+
+	coins, redeemScripts, fundingCoins, spents, err := btc.fundMultiBestEffort(keep+reserves, values, maxFeeRate)
+	if err != nil {
+		return nil, nil, 0, err
+	}
+	if len(coins) == len(values) || !allowSplit {
+		for pt, fc := range fundingCoins {
+			btc.fundingCoins[pt] = fc
+		}
+		btc.node.lockUnspent(false, spents)
+		return coins, redeemScripts, 0, nil
+	}
+
+	return btc.fundMultiWithSplit(keep+reserves, values, splitTxFeeRate, maxFeeRate, splitBuffer)
```

Comment:
> Something feels backwards. If I want a split, and you don't split, that would be wrong.

I believe this has been the behavior since the beginning, at least when it is decided the split tx adds more "baggage" than the split would save by avoiding overlock.  Perhaps the context is different here?  Is the bool even "more optional" in this case?

Reply:
The `MultiTrade` will never lock more the balance.Available - keep` funds. The bot will pass `keep` as the amount that has been reserved by bots trading on other markets. If splitting is not allowed, then `MultiTrade` will only fund   the orders that it is able to without dipping into the other bots' allocation.

Reply:
> @martonp please check my understanding. Ignoring fees, if we want to fund 3 1-lot orders with a lot size of 1, and we have three utxos of 1, 1, and 1000, we would end up locking up all of it because fundMultiBestEffort would succeed without error, resulting in an overlock of 999 and precluding any additional orders until the overlocked order matches and a swap is generated. And that is regardless of what the user has set for fundMultiOptions.Split.

Yes, this is correct, as long as `maxLock` is less than or equal to 1002. I'm thinking maybe the confusion is coming from the fact that the goal if splitting in `FundOrder` is avoiding overlock, but the goal of splitting in `FundMultiOrder` is to be able to fund multiple orders, and `maxLock` is what deals with the overlocking. Maybe splitting in `FundMultiOrder` should be called something else?

Reply:
> I'm thinking maybe the confusion is coming from the fact that the goal if splitting in `FundOrder` is avoiding overlock, but the goal of splitting in `FundMultiOrder` is to be able to fund multiple orders, and `maxLock` is what deals with the overlocking.

Adding to the confusion, the default for `allowSplit` is `btc.useSplitTx()` which is set in configuration and has a known definition. 

How would a caller calculate `maxLock`? They won't know how to calculate fees.

Splitting should really be a default behavior for a bot, we just need to be smarter about how we do it. On what reasoning do you think a bot operator may want to not enable splitting?

Reply:
> Adding to the confusion, the default for allowSplit is btc.useSplitTx() which is set in configuration and has a known definition.

True, I'll remove that default.

> How would a caller calculate maxLock? They won't know how to calculate fees.

`maxLock` is set to the balance that is allocated to the bot that is placing the orders. You can see this logic in https://github.com/decred/dcrdex/pull/2332

> Splitting should really be a default behavior for a bot, we just need to be smarter about how we do it. On what reasoning do you think a bot operator may want to not enable splitting?

I think splitting should be the default, but potentially someone would want to turn on the bot and play around with it without automatically having to spend money.

Reply:
> `maxLock` is set to the balance that is allocated to the bot that is placing the orders. You can see this logic in #2332

Ah. I see `form.MaxLock = c.balanceHandler.botBalance(c.botID, fromAsset)` now.

> .. `maxLock` is what deals with the overlocking. 

If `maxLock` is being used to indicate the balance available to the bot, can it simultaneously be used to limit overlock?

Reply:
The bot code could potentially calculate the amount required for the orders, add a buffer on top, and pass that for max lock, if that's desired. The main thing I was concerned with here is not using the other bots' funds.

Reply:
Another simple approach would be to set maxLock to the smaller of the bot balance or some multiple of the total-value of the multi-trade.  e.g. in the (1,1,1000) UTXO case where the total order value is just 3, if you set maxLock to 9 and avoid insane levels of overlock.

Reply:
Regarding default to split, always doing a split is not great for several reasons, the top being the fees, but also the greatly slowed order submission because server is then waiting on network tx propagation to verify the fresh 0-conf funding UTXO.